fs: support BigInt in fs.*stat and fs.watchFile

joyeecheung opened the pull request:
- fs: support BigInt in fs.*stat and fs.watchFile
  Add the `bigint: true` option to all the `fs.*stat` methods and
  `fs.watchFile`.
- doc: document BigInt support in fs.Stats

I chose to expose the BigInt variant via `bigint: true` in the option object because right now there are no support of a Date class with higher precision - the old Date is Number-based and can lose precision when being constructed out of a 64bit ms value. The plan is that when JS does get a better Date e.g. [Temporal](https://github.com/tc39/proposal-temporal) (see https://github.com/tc39/proposal-bigint/issues/136), we will be able to support `temporal: true`.

The `--harmony-bigint` flag is necessary until v8 6.7 lands.

Fixes: https://github.com/nodejs/node/issues/12115

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

vsemozhetbyt commented on doc/api/fs.md:
```diff
@@ -2693,7 +2748,7 @@ Synchronous rmdir(2). Returns `undefined`.
 on Windows and an `ENOTDIR` error on POSIX.
 
-## fs.stat(path, callback)
+## fs.stat(path[, options], callback)
```
When we change a heading, we should grep all-docs-wide to fix all possible cross-links. It seems there are two more with old hashes:

[deprecations.md#L1037](https://github.com/nodejs/node/blame/985d180855d1d57f64ee213ed72fb460328f53f0/doc/api/deprecations.md#L1037)

[fs.md#L4681](https://github.com/nodejs/node/blame/53822f605d3171530ab820d71e8832f473479c27/doc/api/fs.md#L4681)

BridgeAR commented on test/parallel/test-fs-stat-bigint.js:
```diff
@@ -0,0 +1,145 @@
+  const numStats = await handle.stat();
+  verifyStats(bigintStats, numStats);
+  await handle.close();
+})();
```
These tests would all pass in case e.g. `const handle = await promiseFs.open(fn, 'r');` rejects. In that case it would log an unhandled rejection and that would currently not end the process, so we would not notice anything wrong.

Instead, it would be good to write something like:

```js
async function foo() {
}

Promise.all([
  foo(),
  bar(),
  ...
]).then(common.mustCall());

BridgeAR replied:
There is also a `common` function to throw in case of unhandled rejections.

joyeecheung replied:
I went with `common.crashOnUnhandledRejection()` because it's simple

refack commented:
Can one of the admins verify this patch?

joyeecheung commented:
I think I've addressed all the comments from @vsemozhetbyt and @BridgeAR . This should be ready now that v8 6.7 has landed on master. PTAL @nodejs/fs 

CI: https://ci.nodejs.org/job/node-test-pull-request/15249/

bnoordhuis commented:
Looks like `lib/internal/fs/promises.js` is missing this:
```diff
diff --git a/lib/internal/fs/promises.js b/lib/internal/fs/promises.js
index 43956dae3f..a1670fc65b 100644
--- a/lib/internal/fs/promises.js
+++ b/lib/internal/fs/promises.js
@@ -476,6 +476,7 @@ module.exports = {
   symlink,
   lstat,
   stat,
+  fstat,
   link,
   unlink,
   chmod,
```

joyeecheung commented:
@bnoordhuis `fsPromises.fstat` should already be removed by https://github.com/nodejs/node/pull/20559 , I forgot to update the tests. I'll update it.

joyeecheung commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/15257/

bnoordhuis commented:
Still saying `TypeError: promiseFs.fstat is not a function`.

joyeecheung commented:
https://ci.nodejs.org/job/node-test-pull-request/15260/

Oops, forgot to push

bnoordhuis commented:
Still failing on Windows, I'm afraid: https://ci.nodejs.org/job/node-test-binary-windows/17813/COMPILED_BY=vs2017,RUNNER=win2008r2-vs2017,RUN_SUBSET=2/console

joyeecheung commented:
Green on my Windows laptop now. @bnoordhuis Would you mind take a look at the last commit? I think it should be OK to special case for Windows like that.

Another CI: https://ci.nodejs.org/job/node-test-pull-request/15266/

joyeecheung commented on lib/internal/fs/utils.js:
```diff
@@ -155,6 +155,10 @@ function Stats(
 
 Stats.prototype._checkModeProperty = function(property) {
+  if (isWindows && (property === S_IFIFO || property === S_IFBLK ||
+    property === S_IFSOCK)) {
```
These three are in fact all undefined on Windows, but I figured it looks more readable if I list them like that.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -352,9 +352,15 @@ v8::Local<v8::Value> FillStatsArray(AliasedBuffer<NativeT, V8T>* fields_ptr,
-  return node::FillStatsArray(env->fs_stats_field_array(), s, offset);
+                                          const uv_stat_t* s,
+                                          bool use_bigint = false,
+                                          int offset = 0) {
```
Alignment is off.

bnoordhuis commented on src/node_stat_watcher.h:
```diff
@@ -57,6 +57,7 @@ class StatWatcher : public AsyncWrap {
   bool IsActive();
 
   uv_fs_poll_t* watcher_;
+  bool use_bigint_;
```
`const bool`?

joyeecheung commented:
Test failures seem to be unrelated:

Failures in job https://ci.nodejs.org/job/node-test-commit/18998/ 

#### [alpine-latest-x64](https://ci.nodejs.org/job/node-test-commit-linux/nodes=alpine-latest-x64/19330/console)

<details>
<summary>See failures</summary>

```
not ok 2193 sequential/test-inspector-port-zero-cluster
  ---
  duration_ms: 0.413
  severity: fail
  exitcode: 1
  stack: |-
    Debugger listening on ws://127.0.0.1:39191/3e6681d2-89d9-4b27-989b-baed7a6a650b
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:39192 failed: address already in use
    Assertion failed: !uv__is_closing(handle) (../deps/uv/src/unix/core.c: uv_close: 117)
    Debugger listening on ws://127.0.0.1:39193/708697be-39f0-4229-96b0-354e4b76ccca
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:39194 failed: address already in use
    Assertion failed: !uv__is_closing(handle) (../deps/uv/src/unix/core.c: uv_close: 117)
    assert.js:80
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: code: null, signal: SIGABRT
        at Worker.worker.on.common.mustCall (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-latest-x64/test/sequential/test-inspector-port-zero-cluster.js:20:16)
        at Worker.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-latest-x64/test/common/index.js:443:15)
        at Worker.emit (events.js:182:13)
        at ChildProcess.worker.process.once (internal/cluster/master.js:193:12)
        at Object.onceWrapper (events.js:273:13)
        at ChildProcess.emit (events.js:182:13)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
  ...

```
</details>

#### [fedora-latest-x64](https://ci.nodejs.org/job/node-test-commit-linux/nodes=fedora-latest-x64/19330/console)

<details>
<summary>See failures</summary>

```
error: argument --input/-i: can't open 'test.tap': [Errno 2] No such file or directory: 'test.tap'
POST BUILD TASK : FAILURE
END OF POST BUILD TASK : 0
Recording test results
Notifying upstream projects of job completion
Finished: FAILURE
```
</details>

#### [centos7-arm64-gcc6](https://ci.nodejs.org/job/node-test-commit-arm/nodes=centos7-arm64-gcc6/16622/console)

<details>
<summary>See failures</summary>

```
not ok 2246 sequential/test-inspector-port-zero-cluster
  ---
  duration_ms: 0.332
  severity: fail
  exitcode: 1
  stack: |-
    Debugger listening on ws://127.0.0.1:34395/ed174642-2254-4d6e-83e7-e18fceb7ae24
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:34396 failed: address already in use
    node: ../deps/uv/src/unix/core.c:117: uv_close: Assertion `!(((handle)->flags & (UV_CLOSING | UV_CLOSED)) != 0)' failed.
    Debugger listening on ws://127.0.0.1:34397/8f93d94c-ae7b-4067-806f-f145561dcf5c
    For help, see: https://nodejs.org/en/docs/inspector
    Debugger listening on ws://127.0.0.1:34398/a6c6b38b-7b83-4a86-9897-279ca8d6d6e3
    For help, see: https://nodejs.org/en/docs/inspector
    assert.js:80
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: code: null, signal: SIGABRT
        at Worker.worker.on.common.mustCall (/home/iojs/build/workspace/node-test-commit-arm/nodes/centos7-arm64-gcc6/test/sequential/test-inspector-port-zero-cluster.js:20:16)
        at Worker.<anonymous> (/home/iojs/build/workspace/node-test-commit-arm/nodes/centos7-arm64-gcc6/test/common/index.js:443:15)
        at Worker.emit (events.js:182:13)
        at ChildProcess.worker.process.once (internal/cluster/master.js:193:12)
        at Object.onceWrapper (events.js:273:13)
        at ChildProcess.emit (events.js:182:13)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
  ...

```
</details>

#### [debian7-docker-armv7](https://ci.nodejs.org/job/node-test-commit-arm/nodes=debian7-docker-armv7/16612/): Unknown

joyeecheung commented:
Addressed nits. New CI: https://ci.nodejs.org/job/node-test-pull-request/15277/

joyeecheung commented:
CI is green. Pinging @nodejs/fs again for more reviews.

joyeecheung commented:
I plan to land this later today and try backporting it after https://github.com/nodejs/node/pull/21172 lands.

joyeecheung commented:
Failures seem to be unrelated:

Failures in job https://ci.nodejs.org/job/node-test-commit/19053/ 

#### [centos7-64-gcc6](https://ci.nodejs.org/job/node-test-commit-linux/nodes=centos7-64-gcc6/19385/console)

<details>
<summary>See failures</summary>

```
not ok 2179 sequential/test-inspector-port-zero-cluster
  ---
  duration_ms: 0.280
  severity: fail
  exitcode: 1
  stack: |-
    Debugger listening on ws://127.0.0.1:37752/8c04df0a-cdec-4543-b2fa-2ec22a3495df
    For help, see: https://nodejs.org/en/docs/inspector
    Debugger listening on ws://127.0.0.1:37753/fcae2cf6-2517-4b80-8c02-1341177bc72b
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:37754 failed: address already in use
    node: ../deps/uv/src/unix/core.c:117: uv_close: Assertion `!(((handle)->flags & (UV_CLOSING | UV_CLOSED)) != 0)' failed.
    assert.js:80
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: code: null, signal: SIGABRT
        at Worker.worker.on.common.mustCall (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos7-64-gcc6/test/sequential/test-inspector-port-zero-cluster.js:20:16)
        at Worker.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos7-64-gcc6/test/common/index.js:451:15)
        at Worker.emit (events.js:182:13)
        at ChildProcess.worker.process.once (internal/cluster/master.js:193:12)
        at Object.onceWrapper (events.js:273:13)
        at ChildProcess.emit (events.js:182:13)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
    Debugger listening on ws://127.0.0.1:37755/61ef8c5b-1ab7-41d2-a76e-e0aa9ec6194e
    For help, see: https://nodejs.org/en/docs/inspector
  ...

```
</details>

#### [ubuntu1604_sharedlibs_openssl110_x64](https://ci.nodejs.org/job/node-test-commit-linux-containered/nodes=ubuntu1604_sharedlibs_openssl110_x64/4991/console)

<details>
<summary>See failures</summary>

```
not ok 1961 parallel/test-worker-dns-terminate
  ---
  duration_ms: 0.985
  severity: crashed
  exitcode: -11
  stack: |-
  ...

```
</details>

Failures in job https://ci.nodejs.org/job/node-test-commit/18998/ 

#### [alpine-latest-x64](https://ci.nodejs.org/job/node-test-commit-linux/nodes=alpine-latest-x64/19330/console)

<details>
<summary>See failures</summary>

```
not ok 2193 sequential/test-inspector-port-zero-cluster
  ---
  duration_ms: 0.413
  severity: fail
  exitcode: 1
  stack: |-
    Debugger listening on ws://127.0.0.1:39191/3e6681d2-89d9-4b27-989b-baed7a6a650b
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:39192 failed: address already in use
    Assertion failed: !uv__is_closing(handle) (../deps/uv/src/unix/core.c: uv_close: 117)
    Debugger listening on ws://127.0.0.1:39193/708697be-39f0-4229-96b0-354e4b76ccca
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:39194 failed: address already in use
    Assertion failed: !uv__is_closing(handle) (../deps/uv/src/unix/core.c: uv_close: 117)
    assert.js:80
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: code: null, signal: SIGABRT
        at Worker.worker.on.common.mustCall (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-latest-x64/test/sequential/test-inspector-port-zero-cluster.js:20:16)
        at Worker.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-latest-x64/test/common/index.js:443:15)
        at Worker.emit (events.js:182:13)
        at ChildProcess.worker.process.once (internal/cluster/master.js:193:12)
        at Object.onceWrapper (events.js:273:13)
        at ChildProcess.emit (events.js:182:13)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
  ...

```
</details>

#### [fedora-latest-x64](https://ci.nodejs.org/job/node-test-commit-linux/nodes=fedora-latest-x64/19330/console)

<details>
<summary>See failures</summary>

```
error: argument --input/-i: can't open 'test.tap': [Errno 2] No such file or directory: 'test.tap'
POST BUILD TASK : FAILURE
END OF POST BUILD TASK : 0
Recording test results
Notifying upstream projects of job completion
Finished: FAILURE
```
</details>

#### [centos7-arm64-gcc6](https://ci.nodejs.org/job/node-test-commit-arm/nodes=centos7-arm64-gcc6/16622/console)

<details>
<summary>See failures</summary>

```
not ok 2246 sequential/test-inspector-port-zero-cluster
  ---
  duration_ms: 0.332
  severity: fail
  exitcode: 1
  stack: |-
    Debugger listening on ws://127.0.0.1:34395/ed174642-2254-4d6e-83e7-e18fceb7ae24
    For help, see: https://nodejs.org/en/docs/inspector
    Starting inspector on 127.0.0.1:34396 failed: address already in use
    node: ../deps/uv/src/unix/core.c:117: uv_close: Assertion `!(((handle)->flags & (UV_CLOSING | UV_CLOSED)) != 0)' failed.
    Debugger listening on ws://127.0.0.1:34397/8f93d94c-ae7b-4067-806f-f145561dcf5c
    For help, see: https://nodejs.org/en/docs/inspector
    Debugger listening on ws://127.0.0.1:34398/a6c6b38b-7b83-4a86-9897-279ca8d6d6e3
    For help, see: https://nodejs.org/en/docs/inspector
    assert.js:80
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: code: null, signal: SIGABRT
        at Worker.worker.on.common.mustCall (/home/iojs/build/workspace/node-test-commit-arm/nodes/centos7-arm64-gcc6/test/sequential/test-inspector-port-zero-cluster.js:20:16)
        at Worker.<anonymous> (/home/iojs/build/workspace/node-test-commit-arm/nodes/centos7-arm64-gcc6/test/common/index.js:443:15)
        at Worker.emit (events.js:182:13)
        at ChildProcess.worker.process.once (internal/cluster/master.js:193:12)
        at Object.onceWrapper (events.js:273:13)
        at ChildProcess.emit (events.js:182:13)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
  ...

```
</details>

#### [debian7-docker-armv7](https://ci.nodejs.org/job/node-test-commit-arm/nodes=debian7-docker-armv7/16612/): Unknown

joyeecheung commented:
Landed in af2a104...f54a598, thanks!

joyeecheung commented:
@MylesBorins Can I add backport-requested labels in advance? This should be backported to v10

targos commented:
@joyeecheung 

> This should be backported to v10

It lands cleanly on my side. Unless you know of something that must be manually done for v10.x, I can push it to staging now.

joyeecheung commented:
@targos Yes and yes. Thanks!
